{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c6da3567",
   "metadata": {},
   "source": [
    "# Extract gene IDs with significant alien index scores\n",
    "12B1 and UTEX 2797 predicted proteomes were run through the alien index pipeline. \n",
    "\n",
    "Genes were considered candidates if:\n",
    "\n",
    "* AI > 0.1\n",
    "* total hits in dbs >= 50\n",
    "* if top hit to eukaryote\n",
    "    * hits to other haptophytes >= 5\n",
    "    * hits to dinoflagellates were skipped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d3709be",
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = '../query/alien_index_recipientPRYMNESIUM_ancestralHAPTISTA.txt'\n",
    "ogfile = '../prymnesium_paper/7_Pangenome/1_OrthoFinder/Orthogroups/Orthogroups.txt'\n",
    "nodesfile = '../dbs/nodes.dmp'\n",
    "mergedfile = '../dbs/merged.dmp'\n",
    "\n",
    "# contaminated contigs removed from assembly in v0.4\n",
    "skiplist = ['g2235.t1_12B1', 'g2236.t2_12B1', 'g2237.t1_12B1', 'g2238.t1_12B1', 'g2239.t1_12B1', 'g2240.t1_12B1', 'g2241.t1_12B1', 'g2242.t1_12B1', 'g2243.t1_12B1', 'g6975.t1_12B1', 'g6976.t1_12B1', 'g6977.t1_12B1', 'g6978.t1_12B1', 'g6979.t1_12B1', 'g6980.t1_12B1', 'g6981.t1_12B1', 'g6982.t1_12B1']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a57b711",
   "metadata": {},
   "outputs": [],
   "source": [
    "fi = open(ogfile)\n",
    "\n",
    "ogDict = {}\n",
    "revogDict = {}\n",
    "\n",
    "for line in fi:\n",
    "    genelist = line.rstrip().split()\n",
    "    og = genelist.pop(0).split(':')[0]\n",
    "    \n",
    "    #print(og, genelist)\n",
    "    ogDict[og] = genelist\n",
    "    \n",
    "    for gene in genelist: \n",
    "        revogDict[gene] = og\n",
    "\n",
    "fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b33caaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "parentDict = {}\n",
    "lin = {}\n",
    "\n",
    "nfi = open(nodesfile)\n",
    "for line in nfi:\n",
    "    #print(line)\n",
    "    col = line.rstrip().split('\\t|\\t')\n",
    "    node = col[0]\n",
    "    parent = col[1]\n",
    "    #print(node,parent)\n",
    "    parentDict[node] = parent\n",
    "    \n",
    "nfi.close()\n",
    "\n",
    "mfi = open(mergedfile)\n",
    "for line in mfi:\n",
    "    col = line.rstrip().split('\\t|\\t')\n",
    "    node = col[0]\n",
    "    parent = col[1].split('\\t')[0]  \n",
    "    #print(node,parent)\n",
    "    parentDict[node] = parent\n",
    "mfi.close()\n",
    "\n",
    "def taxdump(taxid):\n",
    "    root = 'no'\n",
    "    taxlist = []\n",
    "    \n",
    "    while root == 'no':\n",
    "        #print(taxid)\n",
    "        if taxid == '':\n",
    "            break\n",
    "        if taxid == '1':\n",
    "            root = 'yes'\n",
    "        taxlist.append(taxid)\n",
    "        \n",
    "        if taxid not in parentDict:\n",
    "            break\n",
    "        \n",
    "        taxid = parentDict[taxid]\n",
    "    \n",
    "    return(taxlist)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f215fcec",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile1 = '../query/candidates_from_other_euks_no_dinos.txt'\n",
    "outfile2 = '../query/candidates_from_bacteria.txt'\n",
    "\n",
    "fi = open(infile)\n",
    "fo1 = open(outfile1, 'w')\n",
    "fo2 = open(outfile2, 'w')\n",
    "\n",
    "candidates = set()\n",
    "\n",
    "for line in fi:\n",
    "    if line[0] == '#':\n",
    "        continue\n",
    "        \n",
    "    col = line.rstrip().split('\\t')\n",
    "    gene = col[0]\n",
    "    if gene in skiplist:\n",
    "        continue\n",
    "        \n",
    "    #print(line)\n",
    "    if col[9] == 'na':\n",
    "        continue\n",
    "        \n",
    "    bestTaxid = col[9].split('-')[1]\n",
    "    bestDom = 'notEuk'\n",
    "    ncbi_lineage = taxdump(bestTaxid)\n",
    "    for taxon_id in ncbi_lineage:\n",
    "        if taxon_id == '2864':\n",
    "            bestDom = 'Dino'\n",
    "            break\n",
    "        if taxon_id == '2759':\n",
    "            bestDom = 'Euk'\n",
    "            break\n",
    "            \n",
    "    ai = float(col[13])\n",
    "    noAncest = int(col[17])\n",
    "    noTotal = col[19]\n",
    "    \n",
    "    if noTotal == '>200':\n",
    "        noTotal = 200\n",
    "    else:\n",
    "        noTotal = int(noTotal)\n",
    "\n",
    "    #print(gene,bestTaxid,bestDom,ai,noAncest,noTotal)\n",
    "    \n",
    "    if noTotal >= 50 and ai > 0.1:\n",
    "        #print(gene,bestTaxid,bestDom,ai,noAncest,noTotal)\n",
    "        if bestDom == 'Euk':\n",
    "            if noAncest > 4:\n",
    "                #print(gene,bestTaxid,bestDom,ai,noAncest,noTotal)\n",
    "                candidates.add(gene)\n",
    "                fo1.write(gene + '\\n')\n",
    "        if bestDom == 'notEuk':\n",
    "            candidates.add(gene)\n",
    "            fo2.write(gene + '\\n')\n",
    "    \n",
    "fi.close()\n",
    "fo1.close()\n",
    "fo2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65a51f08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of AI candidates\n",
    "len(candidates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee83092a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ogCandidates = set()\n",
    "\n",
    "for gene in candidates:\n",
    "    #print(gene)\n",
    "    og = revogDict[gene]\n",
    "    ogCandidates.add(og)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a605286",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 221 AI candidates collapsed into 95 orthogroups\n",
    "len(ogCandidates)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7a4c216",
   "metadata": {},
   "source": [
    "# Combine orthogroup sequences with top hits from AI pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5df762c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53b5df13",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqDict = {}\n",
    "fastafile = '../prymnesium_paper/6_Genome_annotation/5_Functional/fasta/all_proteins.fa.combined'\n",
    "\n",
    "for record in SeqIO.parse(fastafile, 'fasta'):\n",
    "    #print(record.id)\n",
    "    seqDict[record.id] = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513d9cfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "hgtDict = {}\n",
    "fastafiles = '../tree-out-Euk/fasta/*fa'\n",
    "\n",
    "for fastafile in glob.glob(fastafiles):\n",
    "    gene = fastafile.split('/')[-1].split('.fa')[0]\n",
    "    hgtDict[gene] = {}\n",
    "    for record in SeqIO.parse(fastafile, 'fasta'):\n",
    "        #print(gene, record.id)\n",
    "        hgtDict[gene][record.id] = str(record.seq)\n",
    "        \n",
    "        \n",
    "fastafiles = '../tree-out-notEuk/fasta/*fa'\n",
    "\n",
    "for fastafile in glob.glob(fastafiles):\n",
    "    gene = fastafile.split('/')[-1].split('.fa')[0]\n",
    "    hgtDict[gene] = {}\n",
    "    for record in SeqIO.parse(fastafile, 'fasta'):\n",
    "        #print(gene, record.id)\n",
    "        hgtDict[gene][record.id] = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23344974",
   "metadata": {},
   "outputs": [],
   "source": [
    "for og in ogCandidates:\n",
    "    #print(og)\n",
    "    \n",
    "    fastaDict = {}\n",
    "    for gene in ogDict[og]:\n",
    "        #print(gene)\n",
    "        fastaDict[gene] = seqDict[gene]\n",
    "        \n",
    "        if gene in hgtDict:\n",
    "            for hitgene in hgtDict[gene]:\n",
    "                if hitgene.split('-')[0] == 'QUERY':\n",
    "                    continue\n",
    "                \n",
    "                fastaDict[hitgene] = hgtDict[gene][hitgene]\n",
    "        \n",
    "    outfile = '../tree-out/fasta/' + og + '.fa'\n",
    "    fo = open(outfile, 'w')\n",
    "    \n",
    "    for gene in fastaDict:\n",
    "        if gene in hgtDict:\n",
    "            fo.write('>AI-' + gene + '\\n' + fastaDict[gene] + '\\n')\n",
    "        else:\n",
    "            fo.write('>' + gene + '\\n' + fastaDict[gene] + '\\n')\n",
    "    \n",
    "    fo.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fde53f25",
   "metadata": {},
   "source": [
    "# Create supplemental table for HGT1-HGT11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c00d2f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc914e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fastafiles = glob.glob('../tree-out/fasta/*fa')\n",
    "fastas = []\n",
    "for file in fastafiles:\n",
    "    fastas.append(file.split('/')[-1].split('.')[0])\n",
    "    \n",
    "alnfiles = glob.glob('../tree-out/mafft/*aln')\n",
    "alignments = {}\n",
    "for file in alnfiles:\n",
    "    og = file.split('/')[-1].split('.')[0]\n",
    "    \n",
    "    if os.stat(file).st_size == 0:\n",
    "        continue\n",
    "    seq_record = next(SeqIO.parse(file, \"fasta\"))\n",
    "    seqlen = len(str(seq_record.seq))\n",
    "    alignments[og] = seqlen\n",
    "\n",
    "trimfiles = glob.glob('../tree-out/trimal/*trim')\n",
    "trimmed_alignments = {}\n",
    "for file in trimfiles:\n",
    "    og = file.split('/')[-1].split('.')[0]\n",
    "    \n",
    "    if os.stat(file).st_size == 0:\n",
    "        continue\n",
    "    seq_record = next(SeqIO.parse(file, \"fasta\"))\n",
    "    seqlen = len(str(seq_record.seq))\n",
    "    trimmed_alignments[og] = seqlen\n",
    "\n",
    "treefiles = glob.glob('../tree-out/tree/*contree')\n",
    "trees = []\n",
    "for file in treefiles:\n",
    "    trees.append(file.split('/')[-1].split('.')[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d774a6a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile = '../query/alien_index_phylogenies.txt'\n",
    "\n",
    "fo = open(outfile, 'w')\n",
    "fo.write('orthogroup\\talignment length\\ttrimmed align length\\ttree built\\n')\n",
    "for og in fastas:\n",
    "    fo.write(og)\n",
    "    if og in alignments:\n",
    "        fo.write('\\t' + str(alignments[og]))\n",
    "    else:\n",
    "        fo.write('\\t')\n",
    "\n",
    "    if og in trimmed_alignments:\n",
    "        fo.write('\\t' + str(trimmed_alignments[og]))\n",
    "    else:\n",
    "        fo.write('\\t')\n",
    "    \n",
    "    if og in trees:\n",
    "        fo.write('\\tyes')\n",
    "    else:\n",
    "        fo.write('\\tno')\n",
    "\n",
    "    fo.write('\\n')\n",
    "    \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78f35f94",
   "metadata": {},
   "source": [
    "# Create combined phylogeny for OG0000128, OG0000258, OG0000491, OG0024434, OG0024714 and OG0003216\n",
    "\n",
    "These 6 orthogroups are part of a larger gene family. \n",
    "\n",
    "OG0000128, OG0000258, OG0000491, OG0024434, and OG0024714 all contain genes with positive AI scores with phylogenies that support HGT. \n",
    "\n",
    "Genes in OG0003216 do not have a high alien index, but the orthogroup sequences were included as distant homologs in first-pass HGT phylogenies. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cf291e1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "oglist = ['OG0000128', 'OG0000258', 'OG0000491', 'OG0010579', 'OG0024434', 'OG0024714', 'OG0003216']\n",
    "ogfile = '../prymnesium_paper/7_Pangenome/1_OrthoFinder/Orthogroups/Orthogroups.tsv'\n",
    "outfile = '../query/genes_for_HGT1.txt'\n",
    "\n",
    "fi = open(ogfile)\n",
    "fo = open(outfile, 'w')\n",
    "\n",
    "for line in fi:\n",
    "    line = line.rstrip().split('\\t')\n",
    "    og = line[0]\n",
    "    if og in oglist:\n",
    "        bgenes = line[2].split(', ')\n",
    "        ugenes = line[14].split(', ')\n",
    "        for gene in bgenes:\n",
    "            if gene != '':\n",
    "                fo.write(gene + '\\n')\n",
    "        for gene in ugenes:\n",
    "            if gene != '':\n",
    "                fo.write(gene + '\\n')\n",
    "\n",
    "fi.close()\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "53825d15",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f0f45143",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqDict = {}\n",
    "fastafile = '../prymnesium_paper/6_Genome_annotation/5_Functional/fasta/all_proteins.fa.combined'\n",
    "\n",
    "for record in SeqIO.parse(fastafile, 'fasta'):\n",
    "    #print(record.id)\n",
    "    seqDict[record.id] = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2fb440bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "finDict = {}\n",
    "fastafiles = '../tree-out/fasta/g*fa'\n",
    "\n",
    "for fastafile in glob.glob(fastafiles):\n",
    "    for record in SeqIO.parse(fastafile, 'fasta'):\n",
    "        if record.id.split('-')[0] == 'QUERY':\n",
    "            #print(record.id)\n",
    "            record.id = record.id.split('-')[1]\n",
    "        finDict[record.id] = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3117657a",
   "metadata": {},
   "outputs": [],
   "source": [
    "hgtList = []\n",
    "fastafiles = '../tree-out-Euk/fasta/*fa'\n",
    "\n",
    "for fastafile in glob.glob(fastafiles):\n",
    "    gene = fastafile.split('/')[-1].split('.fa')[0]\n",
    "    hgtList.append(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "71dbdde7",
   "metadata": {},
   "outputs": [],
   "source": [
    "oglist = ['OG0000128', 'OG0000258', 'OG0000491', 'OG0010579', 'OG0024434', 'OG0024714', 'OG0003216']\n",
    "ogfile = '../prymnesium_paper/7_Pangenome/1_OrthoFinder/Orthogroups/Orthogroups.tsv'\n",
    "\n",
    "fi = open(ogfile)\n",
    "\n",
    "for line in fi:\n",
    "    line = line.rstrip().split('\\t')\n",
    "    og = line.pop(0)\n",
    "    if og in oglist:\n",
    "        for col in line:\n",
    "            genes = col.split(', ')\n",
    "        for gene in genes:\n",
    "            if gene != '':\n",
    "                finDict[gene] = seqDict[gene]\n",
    "\n",
    "fi.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "91993040",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile = '../tree-out/fasta/HGT1.fa'\n",
    "\n",
    "fo = open(outfile, 'w')\n",
    "\n",
    "for gene in finDict:\n",
    "    if gene in hgtList:\n",
    "        fo.write('>AI-' + gene + '\\n' + finDict[gene] + '\\n')\n",
    "    else:\n",
    "        fo.write('>' + gene + '\\n' + finDict[gene] + '\\n')\n",
    "\n",
    "        \n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d847f326",
   "metadata": {},
   "outputs": [],
   "source": [
    "stopseq = 'XP_023337425.1-88015-Eurytemora_affinis-Metazoa'\n",
    "infile = '../tree-out/mafft/HGT1.aln'\n",
    "outfile = '../tree-out/fasta/HGT1_cropped.fa'\n",
    "\n",
    "fo = open(outfile, 'w')\n",
    "\n",
    "for record in SeqIO.parse(infile, 'fasta'):\n",
    "    #print(record.id)\n",
    "    \n",
    "    sequence = str(record.seq).replace('-' , '')\n",
    "    fo.write('>' + record.id + '\\n' + sequence + '\\n')\n",
    "    \n",
    "    if record.id == stopseq:\n",
    "        break\n",
    "        \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9a62b4c",
   "metadata": {},
   "source": [
    "# Create supplemental table for HGT11 viral insertions genomic region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6f44ff8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "fastafiles = glob.glob('../tree-out/fasta/*fa')\n",
    "fastas = []\n",
    "for file in fastafiles:\n",
    "    fastas.append(file.split('/')[-1].split('.')[0])\n",
    "    \n",
    "alnfiles = glob.glob('../tree-out/mafft/*aln')\n",
    "alignments = {}\n",
    "for file in alnfiles:\n",
    "    og = file.split('/')[-1].split('.')[0]\n",
    "    \n",
    "    if os.stat(file).st_size == 0:\n",
    "        continue\n",
    "    seq_record = next(SeqIO.parse(file, \"fasta\"))\n",
    "    seqlen = len(str(seq_record.seq))\n",
    "    alignments[og] = seqlen\n",
    "\n",
    "trimfiles = glob.glob('../tree-out/trimal/*trim')\n",
    "trimmed_alignments = {}\n",
    "for file in trimfiles:\n",
    "    og = file.split('/')[-1].split('.')[0]\n",
    "    \n",
    "    if os.stat(file).st_size == 0:\n",
    "        continue\n",
    "    seq_record = next(SeqIO.parse(file, \"fasta\"))\n",
    "    seqlen = len(str(seq_record.seq))\n",
    "    trimmed_alignments[og] = seqlen\n",
    "\n",
    "treefiles = glob.glob('../tree-out/tree/*contree')\n",
    "trees = []\n",
    "for file in treefiles:\n",
    "    trees.append(file.split('/')[-1].split('.')[0])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5dbe23b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile = '../query/HGT11_region_phylogenies.txt'\n",
    "\n",
    "fo = open(outfile, 'w')\n",
    "fo.write('orthogroup\\talignment length\\ttrimmed align length\\ttree built\\n')\n",
    "for og in fastas:\n",
    "    fo.write(og)\n",
    "    if og in alignments:\n",
    "        fo.write('\\t' + str(alignments[og]))\n",
    "    else:\n",
    "        fo.write('\\t')\n",
    "\n",
    "    if og in trimmed_alignments:\n",
    "        fo.write('\\t' + str(trimmed_alignments[og]))\n",
    "    else:\n",
    "        fo.write('\\t')\n",
    "    \n",
    "    if og in trees:\n",
    "        fo.write('\\tyes')\n",
    "    else:\n",
    "        fo.write('\\tno')\n",
    "\n",
    "    fo.write('\\n')\n",
    "    \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dfc25a1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
