{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quantify subgenome phylogenetic signal in UTEX2797\n",
    "Plot signal as heatmap in circos format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "from matplotlib import pyplot as plt\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parameters\n",
    "interval_size = 500000\n",
    "support_threshold = 50\n",
    "min_band_size = 15\n",
    "\n",
    "# infiles\n",
    "faifile = '../../figshare/scaffolded_assemblies/UTEX2797_scaffolds_v1.fasta.fai'\n",
    "gfffile =  '../../figshare/annotation/genes_scaffolded_assembly/UTEX2797_v1.gff'\n",
    "synmapfile = '../../figshare/coge/UTEX2797vself/64270_64270.CDS-CDS.last.tdd10.cs0.filtered.dag.all.go_D20_g10_A5.aligncoords.Dm0.ma1.txt'\n",
    "\n",
    "orange_bedfile = '../../figshare/orthofinder/Comparative_Genomics_Statistics/UTEX2797_sisrels_orange' + str(support_threshold) + '.bed'\n",
    "pink_bedfile = '../../figshare/orthofinder/Comparative_Genomics_Statistics/UTEX2797_sisrels_pink' + str(support_threshold) + '.bed'\n",
    "\n",
    "# outfiles\n",
    "statsfile = '../../figshare/coge/circos/stats.txt'\n",
    "outdir = '../../figshare/coge/circos/UTEX2797_subgenomes_interval' + str(interval_size) + '_support' + str(support_threshold) + '_minbandsize' + str(min_band_size)\n",
    "configfile = outdir + '/circos.config'\n",
    "karyotypefile = outdir + '/karyotype.txt'\n",
    "linkfile = outdir +'/syntenic_bands.txt'\n",
    "\n",
    "interval_bedfile = outdir + '/intervals_' + str(interval_size) + '.bed'\n",
    "genes_bedfile = outdir + '/all_genes.bed'\n",
    "\n",
    "allgenesint_bedfile = outdir + '/all_genes_intersect_intervals.bed'\n",
    "orangegenesint_bedfile = outdir + '/orange_genes_intersect_intervals.bed'\n",
    "pinkgenesint_bedfile = outdir + '/pink_genes_intersect_intervals.bed'\n",
    "\n",
    "orange_circosfile = outdir + '/orange_subgenome_intervals.txt'\n",
    "pink_circosfile = outdir + '/pink_subgenome_intervals.txt'\n",
    "\n",
    "!mkdir -p {outdir}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create circos karyotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scaffold_length_dict = {}\n",
    "\n",
    "fi = open(faifile)\n",
    "fo = open(karyotypefile, 'w')\n",
    "\n",
    "for line in fi:\n",
    "    line = line.rstrip()\n",
    "\n",
    "    scaffold = line.split('\\t')[0]\n",
    "    length = int(line.split('\\t')[1])\n",
    "    \n",
    "    scaffold_length_dict[scaffold] = length\n",
    "    fo.write('chr\\t-\\t' + scaffold + '\\t' + scaffold.split('-Scaf')[1] + '\\t0\\t' + str(length) + '\\t' + 'black' + '\\n')\n",
    "\n",
    "fi.close()\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Store interval and gene information as bed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fo = open(interval_bedfile, 'w')\n",
    "\n",
    "for scaffold in scaffold_length_dict:\n",
    "    length = scaffold_length_dict[scaffold]\n",
    "    max_intervals = int(length / interval_size) + 1\n",
    "\n",
    "    intervals = range(0,max_intervals)\n",
    "\n",
    "    for interval in intervals:\n",
    "        intervalid = scaffold + '_' + str(interval)\n",
    "        start = (interval_size * interval) + 1\n",
    "        stop = interval_size * (interval + 1)\n",
    "\n",
    "        if stop > length:\n",
    "            stop = length\n",
    "        \n",
    "        fo.write(scaffold + '\\t' + str(start) + '\\t' + str(stop) + '\\t' + intervalid + '\\n')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = ['genes', 'orange0', 'pink0', 'orange50', 'pink50', 'orange90', 'pink90']\n",
    "\n",
    "fi = open(gfffile)\n",
    "fo = open(genes_bedfile, 'w')\n",
    "\n",
    "counts_dict = {} \n",
    "\n",
    "for line in fi:\n",
    "    feature = line.split('\\t')[2]\n",
    "    \n",
    "    if feature != 'gene':\n",
    "        continue\n",
    "\n",
    "    geneid = line.split('\\t')[8].split(';')[0].split('=')[1]\n",
    "    #print(geneid)\n",
    "\n",
    "    scaffold = line.split('\\t')[0]\n",
    "    if scaffold not in counts_dict:\n",
    "        counts_dict[scaffold] = {}\n",
    "        for geneset in header:\n",
    "            counts_dict[scaffold][geneset] = []\n",
    "        \n",
    "    start = int(line.split('\\t')[3])\n",
    "    stop = int(line.split('\\t')[4])\n",
    "\n",
    "    fo.write(scaffold + '\\t' + str(start) + '\\t' + str(stop) + '\\t' + geneid + '\\n')\n",
    "    counts_dict[scaffold]['genes'].append(geneid)\n",
    "    \n",
    "fi.close()\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intervals = BedTool(interval_bedfile)\n",
    "all_genes = BedTool(genes_bedfile)\n",
    "orange_genes = BedTool(orange_bedfile)\n",
    "pink_genes = BedTool(pink_bedfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get counts for supplimental table\n",
    "sisrelfiles = '../../figshare/orthofinder/Comparative_Genomics_Statistics/UTEX2797_sisrels_*.bed'\n",
    "\n",
    "for infile in glob.glob(sisrelfiles):\n",
    "    fi = open(infile)\n",
    "    \n",
    "    geneset = infile.split('_')[-1].split('.')[0]\n",
    "    #print(geneset)\n",
    "    \n",
    "    for line in fi:\n",
    "        if line[0] == '#':\n",
    "            continue\n",
    "        \n",
    "        col = line.rstrip().split('\\t')\n",
    "        scaffold = col[0]\n",
    "                    \n",
    "        geneid = col[3].split('_')[1]\n",
    "        counts_dict[scaffold][geneset].append(geneid)\n",
    "        #print(geneid)\n",
    "    \n",
    "    fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fo = open(statsfile, 'w')\n",
    "fo.write('scaffold')\n",
    "\n",
    "for geneset in header:\n",
    "    fo.write('\\t' + geneset)\n",
    "fo.write('\\n')\n",
    "\n",
    "for i in range(1,67):\n",
    "    scaffold = 'UTEX2797-Scaf' + str(i)\n",
    "    fo.write(scaffold)\n",
    "    for geneset in header:\n",
    "        count = len(counts_dict[scaffold][geneset])\n",
    "        fo.write('\\t' + str(count))\n",
    "    fo.write('\\n')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate intersections between intervals and gene sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fo = open(allgenesint_bedfile, 'w')\n",
    "\n",
    "for interval in intervals.intersect(all_genes, F=0.5, wa=True, wb=True):\n",
    "    #print(interval)\n",
    "    fo.write(str(interval))\n",
    "    \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fo = open(orangegenesint_bedfile, 'w')\n",
    "\n",
    "for interval in intervals.intersect(orange_genes, F=0.5, wa=True, wb=True):\n",
    "    #print(interval)\n",
    "    fo.write(str(interval))\n",
    "    \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fo = open(pinkgenesint_bedfile, 'w')\n",
    "\n",
    "for interval in intervals.intersect(pink_genes, F=0.5, wa=True, wb=True):\n",
    "    #print(interval)\n",
    "    fo.write(str(interval))\n",
    "    \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intervalDict = {}\n",
    "\n",
    "for interval in intervals.intersect(all_genes, c=True, F=0.5):\n",
    "    intervalDict[interval.name] = [interval.start, interval.stop, interval.score]\n",
    "    \n",
    "for interval in intervals.intersect(orange_genes, c=True, F=0.5):\n",
    "    intervalDict[interval.name].append(interval.score)\n",
    "    \n",
    "for interval in intervals.intersect(pink_genes, c=True, F=0.5):\n",
    "    intervalDict[interval.name].append(interval.score)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "orangelist = []\n",
    "pinklist = []\n",
    "\n",
    "fo1 = open(orange_circosfile, 'w')\n",
    "fo2 = open(pink_circosfile, 'w')\n",
    "\n",
    "for interval in intervalDict:\n",
    "    orange_count = float(intervalDict[interval][3])\n",
    "    \n",
    "    orange_per = 0\n",
    "    if float(intervalDict[interval][2]) > 0:\n",
    "        orange_per = orange_count / float(intervalDict[interval][2]) * 100\n",
    "        \n",
    "    pink_count = float(intervalDict[interval][4])\n",
    "    pink_per = 0\n",
    "    if float(intervalDict[interval][2]) > 0:\n",
    "        pink_per = pink_count / float(intervalDict[interval][2]) * 100\n",
    "    \n",
    "    #print(interval, orange_count, orange_per, pink_count, pink_per)\n",
    "    scaffold = interval.split('_')[0]\n",
    "    start = intervalDict[interval][0]\n",
    "    stop = intervalDict[interval][1]\n",
    "    \n",
    "    fo1.write(scaffold + '\\t' + str(start) + '\\t' + str(stop) + '\\t' + str(orange_per) + '\\n')\n",
    "    fo2.write(scaffold + '\\t' + str(start) + '\\t' + str(stop) + '\\t' + str(pink_per) + '\\n')\n",
    "    \n",
    "    orangelist.append(orange_per)\n",
    "    pinklist.append(pink_per)\n",
    "    \n",
    "fo1.close()\n",
    "fo2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('ORANGE RANGE:', min(orangelist), max(orangelist))\n",
    "print('PINK RANGE:', min(pinklist), max(pinklist))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse synteny information from quota alignment results downloaded from CoGe\n",
    "\n",
    "Syntentic blocks identified via DAGChainer \n",
    "Input Format: https://genomevolution.org/wiki/index.php/SynMap#Output_Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fi = open(synmapfile)\n",
    "fo = open(linkfile, 'w')\n",
    "\n",
    "a_scaf = ''\n",
    "b_scaf = ''\n",
    "a_start = 1000000000\n",
    "b_start = 1000000000\n",
    "a_stop = 0\n",
    "b_stop = 0\n",
    "score = 0\n",
    "num_genes = 0\n",
    "\n",
    "score_list = []\n",
    "num_gene_list = []\n",
    "\n",
    "for line in fi:\n",
    "    col = line.rstrip().split('\\t')\n",
    "    \n",
    "    if line.startswith('#'):\n",
    "        \n",
    "        if a_scaf != '':\n",
    "            if num_genes >= min_band_size:\n",
    "                #print(a_scaf, a_stop, a_stop, b_scaf, b_start, b_stop, score, num_genes)\n",
    "                fo.write(a_scaf + '\\t' + str(a_start) + '\\t' + str(a_stop) + '\\t' + b_scaf + '\\t' + str(b_start) + '\\t' + str(b_stop) + '\\n')\n",
    "                score_list.append(score)\n",
    "                num_gene_list.append(num_genes)\n",
    "                \n",
    "            a_scaf = ''\n",
    "            b_scaf = ''\n",
    "            a_start = 1000000000\n",
    "            b_start = 1000000000\n",
    "            a_stop = 0\n",
    "            b_stop = 0\n",
    "            score = 0\n",
    "            num_genes = 0\n",
    "\n",
    "        #score = float(col[1])\n",
    "        #num_genes = int(col[5])\n",
    "        #print(score, num_genes)\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        a_scaf = col[1].split('||')[0]\n",
    "        b_scaf = col[5].split('||')[0]\n",
    "        \n",
    "        if a_start > int(col[1].split('||')[1]):\n",
    "            a_start = int(col[1].split('||')[1])\n",
    "\n",
    "        if b_start > int(col[5].split('||')[1]):\n",
    "            b_start = int(col[5].split('||')[1])\n",
    "\n",
    "        if a_stop < int(col[1].split('||')[2]):\n",
    "            a_stop = int(col[1].split('||')[2])\n",
    "\n",
    "        if b_stop < int(col[5].split('||')[2]):\n",
    "            b_stop = int(col[5].split('||')[2])\n",
    "\n",
    "        num_genes += 1 \n",
    "\n",
    "        score = int(col[9])\n",
    "\n",
    "fi.close()            \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot distribution of scores for syntenic pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(min(score_list), '-', max(score_list))\n",
    "plt.hist(score_list, 50)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plot distribution of gene counts for syntenic pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(min(num_gene_list), '-', max(num_gene_list))\n",
    "plt.hist(num_gene_list, 50)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare circos configuration file\n",
    "\n",
    "To run:\n",
    "```circos -conf circos.conf -outputfile sub_genome_circos.png```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fo = open(configfile, 'w')\n",
    "\n",
    "fo.write('''\n",
    "<<include etc/colors_fonts_patterns.conf>>\n",
    "<<include etc/housekeeping.conf>>\n",
    "<<include colors_fonts_patterns.conf>>\n",
    "\n",
    "karyotype = karyotype.txt\n",
    "\n",
    "show_ticks          = no\n",
    "show_tick_labels    = no\n",
    "\n",
    "chromosomes_units           = 1000000\n",
    "chromosomes_display_default = yes\n",
    "\n",
    "<image>\n",
    "\t<<include etc/image.conf>>\n",
    "</image>\n",
    "\n",
    "\n",
    "<ideogram>\n",
    "\t<spacing>\n",
    "\t\tdefault = 0.007r\n",
    "\t\tbreak   = 0.5r\n",
    "\n",
    "\t\t<pairwise UTEX2797-Scaf66 UTEX2797-Scaf1>\n",
    "\t\t\tspacing = 5r \n",
    "\t\t</pairwise>\n",
    "\n",
    "\t</spacing>\n",
    "\n",
    "\tshow_bands       = no\n",
    "\tradius           = 0.90r\n",
    "\tthickness        = 25p\n",
    "\tfill             = yes\n",
    "\tfill_color       = black\n",
    "\tstroke_thickness = 2\n",
    "\tstroke_color     = black\n",
    "\n",
    "\tshow_label       = yes\n",
    "\tlabel_font       = default\n",
    "\t#label_radius    = 0.95r\n",
    "\tlabel_with_tag   = yes\n",
    "\tlabel_size       = 40\n",
    "\tlabel_parallel   = yes\n",
    "\tlabel_case       = lower\n",
    "\tlabel_radius     = dims(ideogram,radius_outer) + 25p\n",
    "\n",
    "</ideogram>\n",
    "\n",
    "\n",
    "<links>\n",
    "\t<link>\n",
    "\t\tfile          = syntenic_bands.txt\n",
    "\t\tradius        = 0.81r\n",
    "\t\tbezier_radius = 0r\n",
    "\t\tcolor         = grey\n",
    "\t\tthickness     = 2\n",
    "\t\tribbon           = yes\n",
    "\t\tflat             = yes\n",
    "\t\tstroke_color     = grey\n",
    "\t\tstroke_thickness = 3\n",
    "\n",
    "\t\t<rules>\n",
    "\t\t\t<rule>\n",
    "\t\t\t\tcondition = var(intrachr)\n",
    "\t\t\t\tshow = no\n",
    "\t\t\t</rule>\n",
    "\t\t</rules>\n",
    "\t</link>\n",
    "</links>\n",
    "\n",
    "<plots>\n",
    "\n",
    "\t# orange heatmap\n",
    "\t<plot>\n",
    "\t\ttype  = heatmap\n",
    "\t\tfile  = orange_subgenome_intervals.txt\n",
    "\t\tr1    = 0.99r\n",
    "\t\tr0    = 0.90r\n",
    "\t\tcolor = ylorbr-9-seq\n",
    "\t</plot>\n",
    "\n",
    "\t# pink heatmap\n",
    "\t<plot>\n",
    "\t\ttype  = heatmap\n",
    "\t\tfile  = pink_subgenome_intervals.txt\n",
    "\t\tr1    = 0.90r\n",
    "\t\tr0    = 0.81r\n",
    "\t\tcolor = purd-9-seq\n",
    "\t</plot>\n",
    "\n",
    "</plots>\n",
    "''')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### To run Circos:\n",
    "# module load circos\n",
    "# cd {outdir}\n",
    "# circos -conf circos.config -outputfile {outdir}_circos.png"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
